descriptions.go

```diff
@@ -126,11 +131,255 @@ type RequestorInfo struct {
+	// that have no dependencies, and after that across the intermediate nodes whose dependencies
+	// have been put in the result slice in previous iterations.
+	var result []IssueWizardItem                         // to return
+	resultmap := map[CredentialTypeIdentifier]struct{}{} // to keep track of credentials already put in the result slice
+	for i := len(bylevel) - 1; i >= 0; i-- {
+		for id := range bylevel[i] {
+			if _, ok := resultmap[id]; ok {
+				continue
+			}
+			resultmap[id] = struct{}{}
+			if item, present := byID[id]; present {
+				result = append(result, item)
+			} else {
+				current := id // create copy of loop variable to take address of
+				result = append(result, IssueWizardItem{
+					Type:       IssueWizardItemTypeCredential,
+					Credential: &current,
+				})
+			}
+		}
+	}
+
+	result = append(result, contents[len(contents)-skipped:]...)
+	return result, nil
+}
+
+type credentialDependencies map[CredentialTypeIdentifier][]IssueWizardItem
+
+// wizardItemVisit is a recursive function that populates a map containing per level of a tree the
+// (deduplicated) nodes at that level.
+func wizardItemVisit(
+	bylevel map[int]map[CredentialTypeIdentifier]struct{},
+	level int,
+	id CredentialTypeIdentifier,
+	conf *Configuration,
+	deps credentialDependencies,
+	creds map[CredentialTypeIdentifier]struct{},
+) error {
+	if level == 25 {
+		return errors.New("issue wizard: level limit exceeded, probably due to circular dependency")
+	}
+
+	if bylevel[level] == nil {
+		bylevel[level] = map[CredentialTypeIdentifier]struct{}{}
+	}
+	bylevel[level][id] = struct{}{}
+
+	for _, child := range deps.get(id, conf, creds) {
+		if err := wizardItemVisit(bylevel, level+1, *child.Credential, conf, deps, creds); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (deps credentialDependencies) get(id CredentialTypeIdentifier, conf *Configuration, creds map[CredentialTypeIdentifier]struct{}) []IssueWizardItem {
+	if _, present := deps[id]; !present {
+		deps[id] = conf.CredentialTypes[id].Dependencies.WizardContents().Choose(conf, creds)
+	}
+	return deps[id]
+}
+
+// Process the wizard contents given the list of present credentials. Of each disjunction,
+// either the first contained inner conjunction that is satisfied by the credential list is chosen;
+// or if no such conjunction exists in the disjunction, the first conjunction is chosen.
+// The result of doing this for all outer conjunctions is flattened and returned.
+func (contents IssueWizardContents) Choose(conf *Configuration, creds map[CredentialTypeIdentifier]struct{}) []IssueWizardItem {
+
+	var choice []IssueWizardItem
+	for _, discon := range contents {
+		disconSatisfied := false
+		for _, con := range discon {
+			conSatisfied := true
+			for _, item := range con {
+				if item.Credential == nil {
+					conSatisfied = false
+					break
+				}
+				if _, present := creds[*item.Credential]; !present {
+					conSatisfied = false
+					break
+				}
+			}
```

Comment:
The code is such that if one of the options in a disjunction contains an item that does *not* contain a `Credential` field, then that option is never chosen by this algorithm. This is done purposefully, because if it is not known in advance which credential the item will issue, then we cannot compare that credential to the list of already present credentials in order to consider the item as completed.

> The 'go to website y' part will not have a Credential field, so it will not satisfy
This is not true: a `website` item (and also a `session` item) may contain a `Credential` field denoting which credential the item is expected to issue. This is done exactly so that such items may participate in this algorithm. I will include a comment explaining this.

Reply:
Ah okay, I didn't know that other types may also contain a `Credential` field (as was also an issue in another comment). So could you say then that if one item in the disjunction contains a `Credential` field, that all elements in the disjunction should specify a `Credential` field then (or some other way to determine that it is finished)? Because then you can check for this on validation.

Reply:
We might do that, but why? Apart from that we could check that on validation, I don't see a good reason for that; it might exclude use cases that we don't currently foresee.

Reply:
Of course, when we add in the future a second method to determine that an action is finished, then then that should be added and then the validation should be extended, but as long as it isn't supported it seems good to me that `irma scheme verify` also doesn't allow you to make it.